src/Compilers/VisualBasic/Portable/Binding/BasesBeingResolvedBinder.vb

```diff
@@ -20,27 +20,61 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-                                                     Optional basesBeingResolved As ConsList(Of Symbol) = Nothing) As AccessCheckResult
+                                                     Optional basesBeingResolved As BasesBeingResolved = Nothing) As AccessCheckResult
             ' Accessibility checking may involve looking at base types. We need to pass this accessibility
             ' checking code any base classes currently being bound so we don't use those.
 
             ' add basesBeingResolved that was passed in to the ones stored in this binder.
             Dim currentBasesBeingResolved = Me.BasesBeingResolved()
-            If basesBeingResolved IsNot Nothing Then
-                For Each sym In basesBeingResolved
-                    currentBasesBeingResolved = New ConsList(Of Symbol)(sym, currentBasesBeingResolved)
+            If Not basesBeingResolved.IsEmpty Then
+                For Each inheritsBeingResolved In If(basesBeingResolved.InheritsBeingResolvedOpt, ConsList(Of TypeSymbol).Empty)
+                    currentBasesBeingResolved = currentBasesBeingResolved.PrependInheritsBeingResolved(inheritsBeingResolved)
+                Next
+                For Each implementsBeingResolved In If(basesBeingResolved.ImplementsBeingResolvedOpt, ConsList(Of TypeSymbol).Empty)
+                    currentBasesBeingResolved = currentBasesBeingResolved.PrependImplementsBeingResolved(implementsBeingResolved)
                 Next
             End If
 
             Return m_containingBinder.CheckAccessibility(sym, useSiteDiagnostics, accessThroughType, currentBasesBeingResolved)
         End Function
     End Class
 
+    Friend Structure BasesBeingResolved
+        Public ReadOnly InheritsBeingResolvedOpt As ConsList(Of TypeSymbol)
+        Public ReadOnly ImplementsBeingResolvedOpt As ConsList(Of TypeSymbol)
+
+        Public Shared ReadOnly Property Empty As BasesBeingResolved
+            Get
+                Return New BasesBeingResolved()
+            End Get
+        End Property
+
+        Public Sub New(inheritsBeingResolved As ConsList(Of TypeSymbol), implementsBeingResolved As ConsList(Of TypeSymbol))
+            Me.InheritsBeingResolvedOpt = inheritsBeingResolved
+            Me.ImplementsBeingResolvedOpt = implementsBeingResolved
+        End Sub
+
+        Public Function PrependInheritsBeingResolved(symbol As TypeSymbol) As BasesBeingResolved
+            Return New BasesBeingResolved(If(InheritsBeingResolvedOpt, ConsList(Of TypeSymbol).Empty).Prepend(symbol), ImplementsBeingResolvedOpt)
+        End Function
+
+        Public Function PrependImplementsBeingResolved(symbol As TypeSymbol) As BasesBeingResolved
+            Return New BasesBeingResolved(InheritsBeingResolvedOpt, If(ImplementsBeingResolvedOpt, ConsList(Of TypeSymbol).Empty).Prepend(symbol))
+        End Function
+
+        Public ReadOnly Property IsEmpty As Boolean
```

Comment:
>IsEmpty [](start = 33, length = 7)

nit: I'm not sure this property is useful given that the caller still has to check both parts individually and handle `Nothing` case. Maybe we can do without it, or split into two properties (one for inherits and one for implements)?

Reply:
> nit: I'm not sure this property is useful given that the caller still has to check both parts individually and handle Nothing case. Maybe we can do without it, or split into two properties (one for inherits and one for implements)?

It is used only in one place, I'll remove it in another PR.

---
In reply to: [315814272](https://github.com/dotnet/roslyn/pull/38103#discussion_r315814272) [](ancestors = 315814272)